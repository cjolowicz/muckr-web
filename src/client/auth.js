// @flow
import 'isomorphic-fetch';
import { API_URL } from '../constants';

export function createBasicAuthHeader(username: string, password: string) {
  const base64 = btoa(`${username}:${password}`);

  return new Headers({
    Authorization: `Basic ${base64}`,
  });
}

export function fetchToken(username: string, password: string) {
  return fetch(`${API_URL}/tokens`, {
    method: 'POST',
    mode: 'cors',
    headers: createBasicAuthHeader(username, password),
  }).then(response => response.json())
    .then(data => data.token);
}
