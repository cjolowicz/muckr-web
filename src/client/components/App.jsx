// @flow
import React from "react";
import { BrowserRouter, Route, Link } from "react-router-dom";

import { withAuth } from "./Auth";
import { ArtistList } from "./ArtistList";
import { API_USER, API_PASSWORD } from "../../constants";

const ArtistListWithAuth = withAuth(ArtistList);

const Index = () => <h2>Welcome</h2>;
const Artists = () => (
  <ArtistListWithAuth username={API_USER} password={API_PASSWORD} />
);

const Header = () => (
  <nav>
    <ul>
      <li>
        <Link to="/">Home</Link>
      </li>
      <li>
        <Link to="/artists/">Artists</Link>
      </li>
    </ul>
  </nav>
);

export const App = () => (
  <BrowserRouter>
    <div>
      <Header />
      <Route path="/" exact component={Index} />
      <Route path="/artists/" component={Artists} />
    </div>
  </BrowserRouter>
);
