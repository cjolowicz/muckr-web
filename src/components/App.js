// @flow
import React from "react";
import { Route, Switch } from "react-router-dom";
import CssBaseline from "@material-ui/core/CssBaseline";

import ArtistsPage from "./pages/ArtistsPage";
import PrivateRoute from "./PrivateRoute";
import Index from "./pages/Index";
import SignIn from "./pages/SignIn";
import Message from "./Message";
import * as routes from "../routes";

const App = () => (
  <>
    <CssBaseline />
    <Switch>
      <Route path={routes.INDEX} exact component={Index} />
      <Route path={routes.SIGNIN} component={SignIn} />
      <PrivateRoute path={routes.ARTISTS} component={ArtistsPage} />
    </Switch>
    <Message />
  </>
);

export default App;
