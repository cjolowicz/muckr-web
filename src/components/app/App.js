// @flow
import React from "react";
import { Route, Switch } from "react-router-dom";
import CssBaseline from "@material-ui/core/CssBaseline";

import ArtistsPage from "../pages/ArtistsPage";
import PrivateRoute from "../PrivateRoute";
import FrontPage from "../pages/FrontPage";
import SignInPage from "../pages/SignInPage";
import * as routes from "../../routes";

const App = () => (
  <>
    <CssBaseline />
    <Switch>
      <Route path={routes.INDEX} exact component={FrontPage} />
      <Route path={routes.SIGNIN} component={SignInPage} />
      <PrivateRoute path={routes.ARTISTS} component={ArtistsPage} />
    </Switch>
  </>
);

export default App;
