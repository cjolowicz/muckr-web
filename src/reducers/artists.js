// @flow
import { combineReducers } from "redux";

import {
  FETCH_ARTISTS_REQUEST,
  FETCH_ARTISTS_SUCCESS,
  FETCH_ARTISTS_FAILURE
} from "../actions/artist";
import type { Action } from "../actions";
import type { Artist } from "../api/artist";
import type { FetchError } from "../api/error";

export type State = {
  isFetching: boolean,
  artists: ?Array<Artist>,
  error: ?FetchError
};

export const initialState: State = {
  isFetching: false,
  artists: null,
  error: null
};

function isFetching(state = initialState.isFetching, action: Action) {
  switch (action.type) {
    case FETCH_ARTISTS_REQUEST:
      return true;
    case FETCH_ARTISTS_SUCCESS:
      return false;
    case FETCH_ARTISTS_FAILURE:
      return false;
    default:
      return state;
  }
}

function artists(state = initialState.artists, action: Action) {
  switch (action.type) {
    case FETCH_ARTISTS_SUCCESS:
      return action.artists;
    case FETCH_ARTISTS_FAILURE:
      return null;
    default:
      return state;
  }
}

function error(state = initialState.error, action: Action) {
  switch (action.type) {
    case FETCH_ARTISTS_REQUEST:
    case FETCH_ARTISTS_SUCCESS:
      return null;
    case FETCH_ARTISTS_FAILURE:
      return action.error;
    default:
      return state;
  }
}

export const isFetchingArtists = (state: State) => state.isFetching;

export const getArtists = (state: State) => state.artists;

export const getArtistsError = (state: State) => state.error;

export default combineReducers<Object, Action>({ isFetching, artists, error });
