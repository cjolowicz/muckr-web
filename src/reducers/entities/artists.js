// @flow
import {
  FETCH_ARTISTS_SUCCESS,
  CREATE_ARTIST_SUCCESS,
  REMOVE_ARTIST_SUCCESS,
  UPDATE_ARTIST_SUCCESS
} from "../../actions/artist";
import type { Action } from "../../actions";
import type { Artist } from "../../api/artist";
import { addById, indexById, removeById } from "../../utils";

export type State = { [number]: Artist };

export const initialState: State = {};

export default function artists(state: State = initialState, action: Action) {
  switch (action.type) {
    case FETCH_ARTISTS_SUCCESS:
      return indexById(action.artists);

    case CREATE_ARTIST_SUCCESS:
      return addById(state, action.artist);

    case REMOVE_ARTIST_SUCCESS:
      return removeById(state, action.id);

    case UPDATE_ARTIST_SUCCESS:
      return addById(state, action.artist);

    default:
      return state;
  }
}
