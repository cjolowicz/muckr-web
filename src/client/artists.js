// @flow
import axios from 'axios';
import { API_URL } from '../constants';
import { createTokenAuthHeader } from './auth';

export function fetchArtists(token: string) {
  return axios.get(`${API_URL}/artists`, {
    headers: createTokenAuthHeader(token),
  }).then(response => response.data);
}
