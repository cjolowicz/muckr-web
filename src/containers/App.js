// @flow
import { connect } from "react-redux";

import App from "../components/App";
import { closeNavigation } from "../actions";

const mapStateToProps = state => ({
  navigationOpen: state.navigation.open
});

export default connect(
  mapStateToProps,
  { closeNavigation }
)(App);
