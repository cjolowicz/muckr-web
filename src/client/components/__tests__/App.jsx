// @flow
import React from "react";
import { mount } from "enzyme";

import { App } from "../App";
import * as artists from "../../artists";
import * as auth from "../../auth";
import { mock } from "../../../test-utils";

describe("renderApp", () => {
  const token =
    "a7e743fb7ca1c7c883312b6abb0c99131e4ff210e0730d2f1ee4fad87e514ea2";
  const artist = { id: 1, name: "Artist" };
  const authPromise = Promise.resolve(token);
  const artistsPromise = Promise.resolve([artist]);

  beforeAll(() => {
    jest.spyOn(auth, "fetchToken").mockReturnValue(authPromise);
    jest.spyOn(artists, "fetchArtists").mockReturnValue(artistsPromise);
  });

  afterAll(() => {
    mock(artists.fetchArtists).mockRestore();
    mock(auth.fetchToken).mockRestore();
  });

  it("renders", () => {
    const wrapper = mount(<App />);
    return Promise.resolve().then(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
