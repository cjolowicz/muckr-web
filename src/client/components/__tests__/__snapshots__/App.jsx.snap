// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderApp renders 1`] = `
<App>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <div>
        <Header>
          <nav>
            <ul>
              <li>
                <Link
                  replace={false}
                  to="/"
                >
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    Home
                  </a>
                </Link>
              </li>
              <li>
                <Link
                  replace={false}
                  to="/artists/"
                >
                  <a
                    href="/artists/"
                    onClick={[Function]}
                  >
                    Artists
                  </a>
                </Link>
              </li>
            </ul>
          </nav>
        </Header>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        >
          <Index
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
          >
            <h2>
              Welcome
            </h2>
          </Index>
        </Route>
        <Route
          component={[Function]}
          path="/artists/"
        />
      </div>
    </Router>
  </BrowserRouter>
</App>
`;
